# Benchmark Comparison: DuckDB vs Pandas 

**Author:** Anurag Kumar Pal
<br>
**Email:** iampalanurag@gmail.com  
**GitHub:** [Anurag-Kumar-Pal](https://github.com/Anurag-Kumar-Pal)  
**Date:** 2026-02-10  
**Description:** Benchmarkig Runtime for DuckDB and Pandas.

---

## Table of Contents
1. [Importing the Libraries](#Importing-the-Libraries)
2. [Data Wrangling using the Pandas Library](#Data-Wrangling-using-the-Pandas-Library)
3. [Data Wrangling using the DuckDB Library](#Data-Wrangling-using-the-DuckDB-Library)
4. [Results](#Results)
5. [Conclusions](#Conclusions)

---

### Importing the Libraries

In [1]:
import pandas as pd
import duckdb as db
import time

### Data Wrangling using the Pandas Library

##### A. Loading the Datasets

In [2]:
%%time

# Datase File for 50k records

pdf_50k_rows = pd.read_csv(r"C:\Users\heyit\Desktop\Jupyter Notebooks\PB Notebooks\AKP_Tech_Playground\DuckDB vs Pandas\Input_Files\50K_Rows.csv")

CPU times: total: 281 ms
Wall time: 279 ms


In [3]:
%%time

# Datase File for 100k records

pdf_100k_rows = pd.read_csv(r"C:\Users\heyit\Desktop\Jupyter Notebooks\PB Notebooks\AKP_Tech_Playground\DuckDB vs Pandas\Input_Files\100K_Rows.csv")

CPU times: total: 719 ms
Wall time: 767 ms


In [4]:
%%time

# Datase File for 500k records

pdf_500k_rows = pd.read_csv(r"C:\Users\heyit\Desktop\Jupyter Notebooks\PB Notebooks\AKP_Tech_Playground\DuckDB vs Pandas\Input_Files\500K_Rows.csv")

CPU times: total: 5.95 s
Wall time: 6.16 s


In [5]:
%%time

# Datase File for 1.5M records

pdf_1point5m_rows = pd.read_csv(r"C:\Users\heyit\Desktop\Jupyter Notebooks\PB Notebooks\AKP_Tech_Playground\DuckDB vs Pandas\Input_Files\1.5M_Rows.csv")

CPU times: total: 8.31 s
Wall time: 8.75 s


In [6]:
%%time

# Datase File for 4M records

pdf_4m_rows = pd.read_csv(r"C:\Users\heyit\Desktop\Jupyter Notebooks\PB Notebooks\AKP_Tech_Playground\DuckDB vs Pandas\Input_Files\4M_Rows.csv")

CPU times: total: 1min 35s
Wall time: 1min 37s


##### B. Checking the # of Rows and Columns for each Dataset

In [7]:
%%time
# Record Count for 50k records

pdf_50k_rows.shape

CPU times: total: 0 ns
Wall time: 0 ns


(48895, 16)

In [8]:
%%time
# Record Count for 100k records

pdf_100k_rows.shape

CPU times: total: 0 ns
Wall time: 0 ns


(114000, 21)

In [9]:
%%time
# Record Count for 500k records

pdf_500k_rows.shape

CPU times: total: 0 ns
Wall time: 0 ns


(568454, 10)

In [10]:
%%time
# Record Count for 1.5M records

pdf_1point5m_rows.shape

CPU times: total: 0 ns
Wall time: 0 ns


(1444963, 11)

In [11]:
%%time
# Record Count for 4M records

pdf_4m_rows.shape

CPU times: total: 0 ns
Wall time: 0 ns


(3906160, 21)

##### C.  Checking the "Sort" Command on Dataset

In [12]:
%%time
# Sorting Data for 50k records

pdf_50k_rows.sort_values("latitude").head(5)

CPU times: total: 15.6 ms
Wall time: 36.1 ms


id                                             name    host_id  \
14119  10830083                Beautiful well kept private home!   56078939   
46919  35489384                                   Cozy Apartment  236186921   
15278  12230928                   Villa DiGioia visit NYC via SI   65806798   
1424     639199  Beautiful 4BR/4BA Home, Staten Island, NY City.    1483081   
23460  18997371                                     Cozy Getaway   90104417   

       host_name neighbourhood_group neighbourhood  latitude  longitude  \
14119       Tony       Staten Island   Tottenville  40.49979  -74.24084   
46919      Iveth       Staten Island   Tottenville  40.50641  -74.23059   
15278  Michael J       Staten Island   Tottenville  40.50708  -74.24285   
1424      Marina       Staten Island   Tottenville  40.50868  -74.23986   
23460     Sueann       Staten Island   Tottenville  40.50873  -74.23914   

             room_type  price  minimum_nights  number_of_reviews last_review  \
14119     Private room    110               2                  0         NaN   
46919  Entire home/apt     75               1                  1  2019-06-28   
15278     Private room    100               2                  0         NaN   
1424   Entire home/apt    299               3                 59  2019-07-08   
23460  Entire home/apt     85               2                 49  2019-07-01   

       reviews_per_month  calculated_host_listings_count  availability_365  
14119                NaN                               1               364  
46919               1.00                               1               299  
15278                NaN                               1               365  
1424                0.82                               1               245  
23460               2.08                               2               159

In [13]:
%%time
# Sorting Data for 100k records

pdf_100k_rows.sort_values("duration_ms").head(5)

CPU times: total: 62.5 ms
Wall time: 51.4 ms


Unnamed: 0                track_id  \
65900       65900  1kR4gIb7nGxHPI3D2ifs59   
59310       59310  6hsyfegVY5yklJneM40mWi   
59812       59812  38Ogh3rsHba83kXx13gbKs   
59775       59775  1HVjSh7scH1PaPiLjy2LEu   
16856       16856  5YKCM3jbJ8lqUXUwfU7KwZ   

                                      artists  \
65900                                     NaN   
59310                              Leila Bela   
59812                              Leila Bela   
59775          Leila Bela;Leila's Opera Class   
16856  Wolfgang Amadeus Mozart;Ingrid Haebler   

                               album_name  \
65900                                 NaN   
59310                         Angra Manyu   
59812                         Angra Manyu   
59775                         Angra Manyu   
16856  Mozart: The Complete Piano Sonatas   

                                              track_name  popularity  \
65900                                                NaN           0   
59310                             The Exorsism Begins...           0   
59812                                                V-4           0   
59775  Screams for a Finale! (feat. Leila's Opera Class)           0   
16856                          Andante in C Major, K. 1a           0   

       duration_ms  explicit  danceability  energy  ...  loudness  mode  \
65900            0     False         0.501  0.5830  ...    -9.460     0   
59310         8586     False         0.000  0.0400  ...   -29.714     0   
59812        13386     False         0.000  0.2240  ...   -22.196     1   
59775        15800     False         0.251  0.5080  ...   -10.564     0   
16856        17453     False         0.467  0.0301  ...   -28.518     0   

       speechiness  acousticness  instrumentalness  liveness  valence  \
65900       0.0605         0.690           0.00396    0.0747    0.734   
59310       0.0000         0.928           0.95600    0.1150    0.000   
59812       0.0000         0.970           0.00000    0.9070    0.000   
59775       0.3160         0.969           0.99900    0.9520    0.000   
16856       0.0428         0.995           0.90000    0.1240    0.000   

         tempo  time_signature  track_genre  
65900  138.391               4        k-pop  
59310    0.000               0      iranian  
59812    0.000               0      iranian  
59775  184.051               3      iranian  
16856   84.375               4    classical  

[5 rows x 21 columns]

In [14]:
%%time
# Sorting Data for 500k records

pdf_500k_rows.sort_values("Time").head(5)

CPU times: total: 250 ms
Wall time: 249 ms


Id   ProductId          UserId         ProfileName  \
150523  150524     6641040   ACITT7DI6IDDL     shari zychinski   
150500  150501     6641040   AJ46FKXOVC7NR  Nicholas A Mesiano   
451855  451856  B00004CXX9   AIUWLEQ1ADEG5    Elizabeth Medina   
230284  230285  B00004RYGX  A344SMIA5JECGM     Vincent P. Ross   
451877  451878  B00004CXX9  A344SMIA5JECGM     Vincent P. Ross   

        HelpfulnessNumerator  HelpfulnessDenominator  Score       Time  \
150523                     0                       0      5  939340800   
150500                     2                       2      5  940809600   
451855                     0                       0      5  944092800   
230284                     1                       2      5  944438400   
451877                     1                       2      5  944438400   

                                                  Summary  \
150523                          EVERY book is educational   
150500  This whole series is great way to spend time w...   
451855                               Entertainingl Funny!   
230284                            A modern day fairy tale   
451877                            A modern day fairy tale   

                                                     Text  
150523  this witty little book makes my son laugh at l...  
150500  I can remember seeing the show when it aired o...  
451855  Beetlejuice is a well written movie ..... ever...  
230284  A twist of rumplestiskin captured on film, sta...  
451877  A twist of rumplestiskin captured on film, sta...

In [15]:
%%time
# Sorting Data for 1.5M records

pdf_1point5m_rows.sort_values("creationDate").head(5)

CPU times: total: 1.89 s
Wall time: 2.17 s


id  reviewId creationDate       criticName  \
155743        ballad_of_aj_weberman   1914317   1800-01-01   Jennie Kermode   
168836                lucky_country   1904549   1800-01-01  Thomas Caldwell   
278238   the_definition_of_insanity   1905309   1800-01-01       Joe Lozito   
947162                        juice   1897051   1800-01-01  Owen Gleiberman   
1060329              heartless-2009   1917013   1800-01-01        Tim Robey   

         isTopCritic originalScore reviewState         publicatioName  \
155743         False           5/5       fresh           Eye for Film   
168836         False           2/5      rotten         Cinema Autopsy   
278238         False         3.5/4       fresh  Big Picture Big Sound   
947162          True            B+       fresh   Entertainment Weekly   
1060329         True           3/5       fresh   Daily Telegraph (UK)   

                                                reviewText scoreSentiment  \
155743                                                 NaN       POSITIVE   
168836                                                 NaN       NEGATIVE   
278238   I don't know how autobiographical this film is...       POSITIVE   
947162   Coming out from behind Spike Lee's camera, Ern...       POSITIVE   
1060329  It's exciting to see a British horror film wit...       POSITIVE   

                                                 reviewUrl  
155743     http://www.eyeforfilm.co.uk/reviews.php?id=8266  
168836   http://blog.cinemaautopsy.com/2009/07/14/film-...  
278238   http://www.bigpicturebigsound.com/The-Definiti...  
947162      http://www.ew.com/ew/article/0,,309271,00.html  
1060329  http://www.telegraph.co.uk/culture/film/filmre...

In [16]:
%%time
# Sorting Data for 4M records

pdf_4m_rows.sort_values("followers_count").head(5)

CPU times: total: 7.95 s
Wall time: 8.23 s


id                       name             universal_name  \
445842   4057402           Mccormack Photos           mccormack-photos   
451480   4060992                Gourd Music                gourd-music   
1230725  4584688      Strayhorn Photography      strayhorn-photography   
1723093  3778148  Soluciones de Iluminación  soluciones-de-iluminación   
1230721  4584680     Little Company Inc The     little-company-inc-the   

                                               description  \
445842           Denver band photography and album design.   
451480                                                 NaN   
1230725                                                NaN   
1723093                                                NaN   
1230721  We bring together problem solvers, strategic t...   

                                              linkedin_url  \
445842   https://www.linkedin.com/company/mccormack-pho...   
451480       https://www.linkedin.com/company/gourd-music/   
1230725  https://www.linkedin.com/company/strayhorn-pho...   
1723093  https://www.linkedin.com/showcase/soluciones-d...   
1230721  https://www.linkedin.com/company/little-compan...   

                             website_url  followers_count  \
445842   http://www.mccormickphotos.net/              0.0   
451480             http://www.gourd.com/              0.0   
1230725                              NaN              0.0   
1723093                              NaN              0.0   
1230721         http://www.littleco.com/              0.0   

         associated_members_count         verification founded_on  ...  \
445842                        1.0  {"verified": false}        NaN  ...   
451480                        1.0  {"verified": false}        NaN  ...   
1230725                       1.0  {"verified": false}        NaN  ...   
1723093                       NaN  {"verified": false}        NaN  ...   
1230721                       0.0  {"verified": false}        NaN  ...   

                                         location_branches logo_url  \
445842   [{"name": "Jersey City", "address": {"city": "...      NaN   
451480   [{"name": "Felton", "address": {"city": "Felto...      NaN   
1230725  [{"name": "Humboldt", "address": {"city": "Hum...      NaN   
1723093                                                 []      NaN   
1230721  [{"name": "Prior Lake", "address": {"city": "P...      NaN   

        specialities              industry hashtags funding_info  \
445842           NaN           Photography      NaN          NaN   
451480           NaN                 Music      NaN          NaN   
1230725          NaN           Photography      NaN          NaN   
1723093          NaN                   NaN      NaN          NaN   
1230721          NaN  Advertising Services      NaN          NaN   

                          __created_at                   __updated_at  \
445842   2025-06-22 06:24:31.712114+00  2025-06-22 06:24:31.712114+00   
451480   2025-06-22 06:33:21.873146+00  2025-06-22 06:33:21.873146+00   
1230725  2025-06-23 06:11:40.249238+00  2025-06-23 06:11:40.249238+00   
1723093  2025-06-24 05:28:34.774188+00  2025-06-24 05:28:34.774188+00   
1230721  2025-06-23 06:11:40.249238+00  2025-06-23 06:11:40.249238+00   

        claimable company_type  
445842          t      company  
451480          t      company  
1230725         t      company  
1723093         f     showcase  
1230721         t      company  

[5 rows x 21 columns]

##### D.  Checking the "Aggregation" Command on Dataset

In [17]:
%%time
# Aggregating Data for 50k records

pdf_50k_rows.groupby("room_type")["price"].mean()

CPU times: total: 0 ns
Wall time: 12.7 ms


room_type
Entire home/apt    211.794246
Private room        89.780973
Shared room         70.127586
Name: price, dtype: float64

In [18]:
%%time
# Aggregating Data for 100k records

pdf_100k_rows.groupby("track_genre")["duration_ms"].mean()

CPU times: total: 31.2 ms
Wall time: 12.7 ms


track_genre
acoustic       214896.957
afrobeat       248412.791
alt-rock       235455.907
alternative    222016.180
ambient        237059.038
                  ...    
techno         312311.477
trance         269007.478
trip-hop       274954.026
turkish        219529.010
world-music    297195.622
Name: duration_ms, Length: 114, dtype: float64

In [19]:
%%time
# Aggregating Data for 500k records

pdf_500k_rows.groupby("Score")["Time"].mean()

CPU times: total: 0 ns
Wall time: 32.8 ms


Score
1    1.303159e+09
2    1.301131e+09
3    1.300126e+09
4    1.296722e+09
5    1.294306e+09
Name: Time, dtype: float64

In [20]:
%%time
# Aggregating Data for 1.5M records

pdf_1point5m_rows.groupby("reviewState")["creationDate"].max()

CPU times: total: 188 ms
Wall time: 200 ms


reviewState
fresh     2023-04-08
rotten    2023-04-08
Name: creationDate, dtype: object

In [21]:
%%time
# Aggregating Data for 4M records

pdf_4m_rows.groupby("industry")["associated_members_count"].mean()

CPU times: total: 422 ms
Wall time: 433 ms


industry
Abrasives and Nonmetallic Minerals Manufacturing    62.250000
Accessible Architecture and Design                   7.714286
Accommodation and Food Services                     54.800000
Accounting                                          29.705513
Administration of Justice                           18.807198
                                                      ...    
Wireless Services                                    8.346405
Wood Product Manufacturing                          47.346154
Writing & Editing                                    8.826317
Writing and Editing                                  3.637877
Zoos and Botanical Gardens                          60.875000
Name: associated_members_count, Length: 523, dtype: float64

##### E.  Checking the "Row-Wise Transformation" Command on Dataset

In [22]:
%%time
# Row-wise Transformation for 50k records

pdf_50k_rows["price_converted"] = pdf_50k_rows["price"].apply(lambda x: x * 1.18)

CPU times: total: 15.6 ms
Wall time: 16.7 ms


In [23]:
%%time
# Row-wise Transformation for 100k records

pdf_100k_rows["duration_ms_converted"] = pdf_100k_rows["duration_ms"].apply(lambda x: x * 1.18)

CPU times: total: 31.2 ms
Wall time: 39.9 ms


In [24]:
%%time
# Row-wise Transformation for 500k records

pdf_500k_rows["Time_converted"] = pdf_500k_rows["Time"].apply(lambda x: x * 1.18)

CPU times: total: 141 ms
Wall time: 136 ms


In [25]:
%%time
# Row-wise Transformation for 1.5M records

pdf_1point5m_rows["reviewId_converted"] = pdf_1point5m_rows["reviewId"].apply(lambda x: x * 1.18)

CPU times: total: 312 ms
Wall time: 328 ms


In [26]:
%%time
# Row-wise Transformation for 4M records

pdf_4m_rows["associated_members_count_converted"] = pdf_4m_rows["associated_members_count"].apply(lambda x: x * 1.18)

CPU times: total: 750 ms
Wall time: 760 ms


### Data Wrangling using the DuckDB Library

##### A. Loading the Datasets

In [27]:
# DuckDB command to read a dataset for 50k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM read_csv_auto(
            'C:\\Users\\heyit\\Desktop\\Jupyter Notebooks\\PB Notebooks\\AKP_Tech_Playground\\DuckDB vs Pandas\\Input_Files\\50K_Rows.csv')
        """

# This prepares the query but doesn't execute it
dbf_50k_rows = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
db_count_records = len(dbf_50k_rows.to_df())
action_time_duckdb = time.perf_counter() - start_action

print("Record Count: ", db_count_records)
print("Timings for 50K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

Record Count:  48895
Timings for 50K Rows -->>>
DuckDb Lazy Read Time:  0.15 s
DuckDB Action Time:  0.21 s


In [28]:
# DuckDB command to read a dataset for 100k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM read_csv_auto(
            'C:\\Users\\heyit\\Desktop\\Jupyter Notebooks\\PB Notebooks\\AKP_Tech_Playground\\DuckDB vs Pandas\\Input_Files\\100K_Rows.csv')
        """

# This prepares the query but doesn't execute it
dbf_100k_rows = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
db_count_records = len(dbf_100k_rows.to_df())
action_time_duckdb = time.perf_counter() - start_action

print("Record Count: ", db_count_records)
print("Timings for 100K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

Record Count:  114000
Timings for 100K Rows -->>>
DuckDb Lazy Read Time:  0.10 s
DuckDB Action Time:  0.28 s


In [29]:
# DuckDB command to read a dataset for 500k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM read_csv_auto(
            'C:\\Users\\heyit\\Desktop\\Jupyter Notebooks\\PB Notebooks\\AKP_Tech_Playground\\DuckDB vs Pandas\\Input_Files\\500K_Rows.csv')
        """

# This prepares the query but doesn't execute it
dbf_500k_rows = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
db_count_records = len(dbf_500k_rows.to_df())
action_time_duckdb = time.perf_counter() - start_action

print("Record Count: ", db_count_records)
print("Timings for 500K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

Record Count:  568454
Timings for 500K Rows -->>>
DuckDb Lazy Read Time:  0.11 s
DuckDB Action Time:  1.22 s


In [30]:
# DuckDB command to read a dataset for 1.5M records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM read_csv_auto(
            'C:\\Users\\heyit\\Desktop\\Jupyter Notebooks\\PB Notebooks\\AKP_Tech_Playground\\DuckDB vs Pandas\\Input_Files\\1.5M_Rows.csv')
        """

# This prepares the query but doesn't execute it
dbf_1point5m_rows = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
db_count_records = len(dbf_1point5m_rows.to_df())
action_time_duckdb = time.perf_counter() - start_action

print("Record Count: ", db_count_records)
print("Timings for 1.5M Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

Record Count:  1444963
Timings for 1.5M Rows -->>>
DuckDb Lazy Read Time:  0.11 s
DuckDB Action Time:  3.03 s


In [31]:
# DuckDB command to read a dataset for 4M records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM read_csv_auto(
            'C:\\Users\\heyit\\Desktop\\Jupyter Notebooks\\PB Notebooks\\AKP_Tech_Playground\\DuckDB vs Pandas\\Input_Files\\4M_Rows.csv', parallel=false)
        """

# This prepares the query but doesn't execute it
dbf_4m_rows = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
db_count_records = len(dbf_4m_rows.to_df())
action_time_duckdb = time.perf_counter() - start_action

print("Record Count: ", db_count_records)
print("Timings for 4M Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Record Count:  3906160
Timings for 4M Rows -->>>
DuckDb Lazy Read Time:  0.21 s
DuckDB Action Time:  64.02 s


##### B.  Checking the "Sort" Command on Dataset

In [32]:
# DuckDB command to sort a dataset of 50k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM dbf_50k_rows
        ORDER BY latitude ASC
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_50k_sorted = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_50k_sorted = dbf_50k_sorted.to_df()
print(pdf_50k_sorted.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Sorting 50K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

         id                                             name    host_id  \
0  10830083                Beautiful well kept private home!   56078939   
1  35489384                                   Cozy Apartment  236186921   
2  12230928                   Villa DiGioia visit NYC via SI   65806798   
3    639199  Beautiful 4BR/4BA Home, Staten Island, NY City.    1483081   
4  18997371                                     Cozy Getaway   90104417   

   host_name neighbourhood_group neighbourhood  latitude  longitude  \
0       Tony       Staten Island   Tottenville  40.49979  -74.24084   
1      Iveth       Staten Island   Tottenville  40.50641  -74.23059   
2  Michael J       Staten Island   Tottenville  40.50708  -74.24285   
3     Marina       Staten Island   Tottenville  40.50868  -74.23986   
4     Sueann       Staten Island   Tottenville  40.50873  -74.23914   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    110               2     

In [33]:
# DuckDB command to sort a dataset of 100k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM dbf_100k_rows
        ORDER BY duration_ms ASC
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_100k_sorted = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_100k_sorted = dbf_100k_sorted.to_df()
print(pdf_100k_sorted.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Sorting 100K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

   column00                track_id                                 artists  \
0     65900  1kR4gIb7nGxHPI3D2ifs59                                    None   
1     59310  6hsyfegVY5yklJneM40mWi                              Leila Bela   
2     59812  38Ogh3rsHba83kXx13gbKs                              Leila Bela   
3     59775  1HVjSh7scH1PaPiLjy2LEu          Leila Bela;Leila's Opera Class   
4     16856  5YKCM3jbJ8lqUXUwfU7KwZ  Wolfgang Amadeus Mozart;Ingrid Haebler   

                           album_name  \
0                                None   
1                         Angra Manyu   
2                         Angra Manyu   
3                         Angra Manyu   
4  Mozart: The Complete Piano Sonatas   

                                          track_name  popularity  duration_ms  \
0                                               None           0            0   
1                             The Exorsism Begins...           0         8586   
2                                  

In [34]:
# DuckDB command to sort a dataset of 500k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM dbf_500k_rows
        ORDER BY Time ASC
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_500k_sorted = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_500k_sorted = dbf_500k_sorted.to_df()
print(pdf_500k_sorted.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Sorting 500K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

       Id   ProductId          UserId         ProfileName  \
0  150524     6641040   ACITT7DI6IDDL     shari zychinski   
1  150501     6641040   AJ46FKXOVC7NR  Nicholas A Mesiano   
2  451856  B00004CXX9   AIUWLEQ1ADEG5    Elizabeth Medina   
3  374359  B00004CI84  A344SMIA5JECGM     Vincent P. Ross   
4  451878  B00004CXX9  A344SMIA5JECGM     Vincent P. Ross   

   HelpfulnessNumerator  HelpfulnessDenominator  Score       Time  \
0                     0                       0      5  939340800   
1                     2                       2      5  940809600   
2                     0                       0      5  944092800   
3                     1                       2      5  944438400   
4                     1                       2      5  944438400   

                                             Summary  \
0                          EVERY book is educational   
1  This whole series is great way to spend time w...   
2                               Entertainingl Funn

In [35]:
# DuckDB command to sort a dataset of 1.5M records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM dbf_1point5m_rows
        ORDER BY creationDate ASC
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_1point5m_sorted = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_1point5m_sorted = dbf_1point5m_sorted.to_df()
print(pdf_1point5m_sorted.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Sorting 1.5M Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

                                id  reviewId creationDate        criticName  \
0       1202487-book_of_revelation   1892314   1800-01-01      Mike Barnard   
1  1013485-masque_of_the_red_death   1908664   1800-01-01    Jennie Kermode   
2        the-jolly-boys-last-stand   1906693   1800-01-01  Christopher Null   
3                       we_believe   1901976   1800-01-01    Erik Childress   
4                      accomplices   1914020   1800-01-01       Roger Ebert   

   isTopCritic originalScore reviewState     publicatioName  \
0        False          5/10      rotten   Future Movies UK   
1        False           5/5       fresh       Eye for Film   
2        False           3/5       fresh     Filmcritic.com   
3        False           4/5       fresh    eFilmCritic.com   
4         True           3/4       fresh  Chicago Sun-Times   

                                          reviewText scoreSentiment  \
0                                               None       NEGATIVE   
1   

In [36]:
# DuckDB command to sort a dataset of 4M records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT * FROM dbf_4m_rows
        ORDER BY followers_count ASC
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_4m_sorted = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_4m_sorted = dbf_4m_sorted.to_df()
print(pdf_4m_sorted.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Sorting 4M Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

        id                        name              universal_name  \
0   941149            Sresht Solutions            sresht-solutions   
1  4298371           Acorn Woodworking           acorn-woodworking   
2   941335  Cross Rhoades Technologies  cross-rhoades-technologies   
3  4298386                  Crc Claims                  crc-claims   
4  2652944            PedalMonsters.co            pedalmonsters-co   

                                         description  \
0  Sresht Solutions is a training and development...   
1                                               None   
2  Check out www.cross-rhoades.com to see what we...   
3                                               None   
4                                               None   

                                        linkedin_url  \
0  https://www.linkedin.com/company/sresht-soluti...   
1  https://www.linkedin.com/company/acorn-woodwor...   
2  https://www.linkedin.com/company/cross-rhoades...   
3       https://ww

##### C.  Checking the "Aggregation" Command on Dataset

In [37]:
# DuckDB command to aggregate a dataset of 50k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT room_type, AVG(price) AS avg_price 
        FROM dbf_50k_rows
        GROUP BY room_type
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_50k_aggregated = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_50k_aggregated = dbf_50k_aggregated.to_df()
print(pdf_50k_aggregated.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Aggregating 50K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

         room_type   avg_price
0     Private room   89.780973
1  Entire home/apt  211.794246
2      Shared room   70.127586
Timings for Aggregating 50K Rows -->>>
DuckDb Lazy Read Time:  0.12 s
DuckDB Action Time:  0.19 s


In [38]:
# DuckDB command to aggregate a dataset of 100k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT track_genre, AVG(duration_ms) AS avg_duration_ms
        FROM dbf_100k_rows
        GROUP BY track_genre
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_100k_aggregated = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_100k_aggregated = dbf_100k_aggregated.to_df()
print(pdf_100k_aggregated.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Aggregating 100K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

  track_genre  avg_duration_ms
0       sleep       184058.980
1   synth-pop       244805.907
2     turkish       219529.010
3      j-idol       256122.622
4       piano       203966.541
Timings for Aggregating 100K Rows -->>>
DuckDb Lazy Read Time:  0.12 s
DuckDB Action Time:  0.12 s


In [39]:
# DuckDB command to aggregate a dataset of 500k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT Score, AVG(Time) AS avg_Time
        FROM dbf_500k_rows
        GROUP BY Score
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_500k_aggregated = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_500k_aggregated = dbf_500k_aggregated.to_df()
print(pdf_500k_aggregated.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Aggregating 500K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

   Score      avg_Time
0      4  1.296722e+09
1      3  1.300126e+09
2      5  1.294306e+09
3      2  1.301131e+09
4      1  1.303159e+09
Timings for Aggregating 500K Rows -->>>
DuckDb Lazy Read Time:  0.12 s
DuckDB Action Time:  0.38 s


In [40]:
# DuckDB command to aggregate a dataset of 1.5M records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT reviewState, MAX(creationDate) AS max_creationDate
        FROM dbf_1point5m_rows
        GROUP BY reviewState
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_1point5m_aggregated = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_1point5m_aggregated = dbf_1point5m_aggregated.to_df()
print(pdf_1point5m_aggregated.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Aggregating 1.5M Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

  reviewState max_creationDate
0       fresh       2023-04-08
1      rotten       2023-04-08
Timings for Aggregating 1.5M Rows -->>>
DuckDb Lazy Read Time:  0.13 s
DuckDB Action Time:  0.41 s


In [41]:
# DuckDB command to aggregate a dataset of 4M records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT industry, AVG(associated_members_count) AS avg_associated_members_count
        FROM dbf_4m_rows
        GROUP BY industry
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_4m_aggregated = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_4m_aggregated = dbf_4m_aggregated.to_df()
print(pdf_4m_aggregated.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Aggregating 4M Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

               industry  avg_associated_members_count
0  Alternative Medicine                     15.380152
1    Financial Services                    109.446535
2        Nanotechnology                     26.261307
3   Program Development                      8.139057
4          Philanthropy                     12.868830
Timings for Aggregating 4M Rows -->>>
DuckDb Lazy Read Time:  0.24 s
DuckDB Action Time:  12.65 s


##### D.  Checking the "Row-Wise Transformation" Command on Dataset

In [42]:
# DuckDB command to aggregate a dataset of 50k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT *, price * 1.18 AS price_converted 
        FROM dbf_50k_rows
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_50k_transformed = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_50k_transformed = dbf_50k_transformed.to_df()
print(pdf_50k_transformed.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Transforming 50K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

     id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  20

In [43]:
# DuckDB command to aggregate a dataset of 100k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT *, duration_ms * 1.18 AS duration_ms_converted 
        FROM dbf_100k_rows
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_100k_transformed = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_100k_transformed = dbf_100k_transformed.to_df()
print(pdf_100k_transformed.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Transforming 100K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

   column00                track_id                 artists  \
0         0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1         1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2         2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3         3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4         4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   

                                          album_name  \
0                                             Comedy   
1                                   Ghost (Acoustic)   
2                                     To Begin Again   
3  Crazy Rich Asians (Original Motion Picture Sou...   
4                                            Hold On   

                   track_name  popularity  duration_ms  explicit  \
0                      Comedy          73       230666     False   
1            Ghost - Acoustic          55       149610     False   
2              To Begin Again          57       210826     False   
3  Can't Hel

In [44]:
# DuckDB command to aggregate a dataset of 500k records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT *, Time * 1.18 AS Time_converted 
        FROM dbf_500k_rows
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_500k_transformed = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_500k_transformed = dbf_500k_transformed.to_df()
print(pdf_500k_transformed.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Transforming 500K Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

   Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   

                 Summary                                               Text  \
0  Good Quality Dog Food  I have bought several of the Vitality canned d...   


In [45]:
# DuckDB command to aggregate a dataset of 1.5M records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT *, reviewId * 1.18 AS reviewId_converted 
        FROM dbf_1point5m_rows
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_1point5m_transformed = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_1point5m_transformed = dbf_1point5m_transformed.to_df()
print(pdf_1point5m_transformed.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Transforming 1.5M Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

                                  id  reviewId creationDate       criticName  \
0                            beavers   1145982   2003-05-23  Ivan M. Lincoln   
1                         blood_mask   1636744   2007-06-02    The Foywonder   
2  city_hunter_shinjuku_private_eyes   2590987   2019-05-28     Reuben Baron   
3  city_hunter_shinjuku_private_eyes   2558908   2019-02-14      Matt Schley   
4                 dangerous_men_2015   2504681   2018-08-29        Pat Padua   

   isTopCritic originalScore reviewState                 publicatioName  \
0        False         3.5/4       fresh  Deseret News (Salt Lake City)   
1        False           1/5      rotten                  Dread Central   
2        False          None       fresh                            CBR   
3        False         2.5/5      rotten                    Japan Times   
4        False          None       fresh                          DCist   

                                          reviewText scoreSentiment 

In [46]:
# DuckDB command to aggregate a dataset of 4M records and convert to Pandas Dataframe
# Lazy Read Simulation: Preparing the query
start_lazy = time.perf_counter()

query = """
        SELECT *, associated_members_count * 1.18 AS associated_members_count_converted 
        FROM dbf_4m_rows
        LIMIT 5
        """

# This prepares the query but doesn't execute it
dbf_4m_transformed = db.sql(query)
lazytime_duckdb = time.perf_counter() - start_lazy

# Actual Materialization of the Query
start_action = time.perf_counter()

# Triggering the DuckDB Query
pdf_4m_transformed = dbf_4m_transformed.to_df()
print(pdf_4m_transformed.head())
action_time_duckdb = time.perf_counter() - start_action

print("Timings for Transforming 4M Rows -->>>")
print(f"DuckDb Lazy Read Time: {lazytime_duckdb: .2f} s")
print(f"DuckDB Action Time: {action_time_duckdb: .2f} s")

        id                               name  \
0   940965                 Bow Plumbing Group   
1   940966  BIONUTRIS CONSULTORIA EM NUTRIÇÃO   
2   940967                    Essèntium Group   
3  2652650               studio legale funari   
4  2652651                  DEMA Partners LTD   

                     universal_name  \
0                bow-plumbing-group   
1  bionutris-consultoria-em-nutri-o   
2        ess-ntium-consulting-group   
3              studio-legale-funari   
4                 dema-partners-ltd   

                                         description  \
0  Founded in 1939, Bow Plumbing Group is one of ...   
1  É com muita satisfação que lhe apresentamos BI...   
2  Essèntium Group is a full-service consulting f...   
3                                               None   
4  Dema Partners LTD is a professional group of c...   

                                        linkedin_url  \
0  https://www.linkedin.com/company/bow-plumbing-...   
1  https://www.linked

### Results

<div style="text-align: center; width: fit-content; margin-left: 0;">

| Operation               | Dataset Size | Pandas Wall Time (s) | DuckDB Action Time (s) | Delta (s) [Pandas - DuckDB]|
|:------------------------|:------------:|:--------------------:|:----------------------:|:------------------:|
| Loading                 | 50k rows     | 0.28                 | 0.21                   | 0.07               |
|                         | 100k rows    | 0.77                 | 0.28                   | 0.49               |
|                         | 500k rows    | 6.16                 | 1.22                   | 4.94               |
|                         | 1.5M rows    | 8.75                 | 3.03                   | 5.72               |
|                         | 4M rows      | 97.00                | 64.02                  | 32.98              |
| Sorting                 | 50k rows     | 0.04                 | 0.38                   | -0.34              |
|                         | 100k rows    | 0.05                 | 0.19                   | -0.14              |
|                         | 500k rows    | 0.25                 | 0.41                   | -0.16              |
|                         | 1.5M rows    | 2.17                 | 0.43                   | 1.74               |
|                         | 4M rows      | 8.23                 | 38.80                  | -30.57             |
| Aggregating             | 50k rows     | 0.01                 | 0.19                   | -0.18              |
|                         | 100k rows    | 0.01                 | 0.12                   | -0.11              |
|                         | 500k rows    | 0.03                 | 0.38                   | -0.35              |
|                         | 1.5M rows    | 0.20                 | 0.41                   | -0.21              |
|                         | 4M rows      | 0.43                 | 12.65                  | -12.22             |
| Row-Wise Transformation | 50k rows     | 0.02                 | 0.14                   | -0.12              |
|                         | 100k rows    | 0.04                 | 0.11                   | -0.07              |
|                         | 500k rows    | 0.14                 | 0.15                   | -0.01              |
|                         | 1.5M rows    | 0.33                 | 0.13                   | 0.20               |
|                         | 4M rows      | 0.76                 | 0.24                   | 0.52               |

</div>


### Conclusions

> - Pandas generally offers better performance for smaller datasets (up to around 500k rows) in most operations
> - DuckDB significantly outperforms Pandas when dealing with large datasets (1.5M+ rows), particularly in loading and aggregation tasks, due to its ability to handle larger volumes efficiently
> - For complex analytical queries and large-scale data processing, DuckDB's architectural advantages make it the more efficient choice from a pure runtime perspective